### PR TITLE
Add GPU implementation of pdist

### PR DESCRIPTION
**Pull request by erikbrinkman**
Add the gpu kernel version.

The parallelism I went with performs poorly when there are a large number of vectors, but they're all short, as I don't allocate the thread pool to wrap in that case.

Test Plan
---------
```
python -m unittest test_torch.TestTorch.test_pdist_{empty,scipy} test_nn.TestNN.test_pdist{,_zeros,_empty_row,_empty_col,_cpu_gradgrad_unimplemented,_cuda_gradgrad_unimplemented} test_jit.TestJitGenerated.test_nn_pdist
```

Current performance specs are a little underwhelming, I'm in the process of debugging.

size | torch | torch cuda | scipy
-----|-------|------------|------
16 x 16 | 9.13 µs ± 3.55 µs | 9.86 µs ± 81.5 ns | 15.8 µs ± 1.2 µs
16 x 1024 | 15 µs ± 224 ns | 9.48 µs ± 88.7 ns | 88.7 µs ± 8.83 µs
1024 x 16 | 852 µs ± 6.03 µs | 7.84 ms ± 6.22 µs | 4.7 ms ± 166 µs
1024 x 1024 | 34.1 ms ± 803 µs | 11.5 ms ± 6.24 µs | 273 ms ± 6.7 ms
2048 x 2048 | 261 ms ± 3.5 ms | 77.5 ms ± 41.5 µs | 2.5 s ± 97.6 ms
4096 x 4096 | 2.37 s ± 154 ms | 636 ms ± 2.97 µs | 25.9 s ± 394 ms

### CODE REVIEW

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 1]
```diff
@@ -291,7 +291,7 @@ def add_dim3(kernel_string, cuda_kernel):
         elif c == ")":
             closure -= 1
         elif (c == "," or ind == len(kernel_string) - 1) and closure == 0:
-            arg_locs[count]['end'] = ind
+            arg_locs[count]['end'] = ind + (c != ",")
```
Comment (ezyang) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: You'll want to mention in the commit message that you fixed this, and what exactly you fixed.

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+#include <math.h>
+
+#include "ATen/native/Distance.h"
+
+
+namespace at { namespace native {
+
+namespace {
+
+static const int warp_size = 32;
```
Comment (colesbury) [inline aten/src/ATen/native/cuda/DistanceKernel.cu]: Avoid hard-coding the size of the warp in CUDA files. It's different, for example, on AMD GPUs.  Using a fixed number of threads per block is fine.

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  const scalar_t * const end = start + m;
+  const scalar_t * a = start + threadIdx.x;
+  const scalar_t * b = self + j * m + threadIdx.x;
+  scalar_t agg = 0.0;
+  for (; a < end; a += stride, b += stride) {
+    F::inc(agg, std::abs(*a - *b), p);
+  }
+  
+  // Reduce warps
+  for (int offset = warp_size / 2; offset > 0; offset /= 2) {
```
Comment (colesbury) [inline aten/src/ATen/native/cuda/DistanceKernel.cu]: Within the kernels you can use the built-in `warpSize`

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    return;
+  }
+
+  const int64_t n = result.size(0);
+  int64_t m = self.size(1);
+  const int horiz_per_block = 2 * warp_size;
+  const int vert_per_block = 4;
+  const int horiz_blocks = (m + horiz_per_block * 8 - 1) / (horiz_per_block * 8);
+  const int vert_blocks = (dist.numel() + vert_per_block - 1) / vert_per_block;
+  const dim3 blocks(horiz_blocks, vert_blocks);
```
Comment (colesbury) [inline aten/src/ATen/native/cuda/DistanceKernel.cu]: The predominant convention in PyTorch is to name these `grid` and `block`

**Review by colesbury — APPROVED** [review]
looks good, just a few comments about code style.

In general, we only label pointer and reference types as `const`

File: aten/src/ATen/native/cuda/DistanceKernel.cu (hunk 2 above)
Comment (erikbrinkman) [reply aten/src/ATen/native/cuda/DistanceKernel.cu]: Ahh, I didn't know `warpSize` was built in. I will switch to that.

**Comment by erikbrinkman** [comment]
@colesbury When you're talking about `const`, are you talking about the `const scalar_t`? I feel like this is important for me to know that I'm not modifying it. Do you want me to remove them, or is there a different const object I should change?

File: aten/src/ATen/native/cuda/DistanceKernel.cu (hunk 3 above)
Comment (erikbrinkman) [reply aten/src/ATen/native/cuda/DistanceKernel.cu]: This becomes an issue when I want to reduce across warps. When I try to compile with shared memory size set to to the number of warps (or some function there of) I get the following compiler error: 

```
note: the value of variable "warpSize" here cannot be used as a constant
```

I could just hard code 32 here, or something similar and just assume that the number of warps is going to be less than that, which seems reasonable but a little eeehhhh.

**Pull request closed without merging**